csrc/device_lower/lower2device.cpp

```diff
@@ -339,17 +339,69 @@ namespace {
 IdModelOptions getIdModelOptions(Fusion* fusion) {
   IdModelOptions options;
 
-  // TMA requires TensorIndexer.
   for (auto expr : fusion->exprs()) {
     if (auto ldst = dynamic_cast<LoadStoreOp*>(expr)) {
       if (ldst->opType() == LoadStoreOpType::CpAsyncBulkTensorTile) {
         options.setBuildTensorIndexer(true);
         continue;
       }
-    }
-    if (expr->isA<MmaOp>()) {
+    } else if (expr->isA<MmaOp>()) {
       options.setBuildTensorIndexer(true);
       continue;
+    } else if (auto reshape = dynamic_cast<ViewOp*>(expr)) {
+      // The legacy indexer has an issue when an expand broadcast is
+      // involved in reshape transformations. Enable both tensor and
+      // predicate indexing if found
+
+      auto producer_tv = reshape->in();
+      auto consumer_tv = reshape->out();
+
+      std::unordered_set<IterDomain*> expanded_ids;
+      std::copy_if(
+          producer_tv->getLogicalDomain().begin(),
+          producer_tv->getLogicalDomain().end(),
+          std::inserter(expanded_ids, expanded_ids.end()),
+          [](IterDomain* logical_id) {
+            return logical_id->isBroadcast() && logical_id->hasExpandedExtent();
+          });
+
+      if (expanded_ids.empty()) {
+        continue;
+      }
+
+      // Find corresponding consumer root IDs
+      auto c2p = PairwiseLogicalDomainMap(producer_tv, consumer_tv)
+                     .mapConsumerToProducer();
+      std::unordered_set<Val*> consumer_expanded_root_ids;
+      for (auto consumer_root_id : consumer_tv->getRootDomain()) {
+        auto producer_logical_id = c2p.at(consumer_root_id);
+        if (expanded_ids.count(producer_logical_id)) {
+          consumer_expanded_root_ids.insert(consumer_root_id);
```

Comment:
Do we need the c2p mapping to populate this set? Wouldn't it always be the case that an ID is in `consumer_expanded_root_ids` if and only if it is itself an expanded broadcast, since we should have a bijection between the root domain of the consumer and the logical domain of the producer? I am wonder if we could just loop over the consumer root domain to build `consumer_expanded_root_ids` and avoid building `expanded_ids` entirely.

Reply:
Reshape with expanded IDs turn them non-broadcast IDs. For example, here's the producer and consumer of the reshape of the repro test case:

```
T4_l_float[ iblockIdx.x60{256000 ex ( ceilDiv(( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) ), 128) )}, iUS59{1}, iS57{4}, ithreadIdx.x61{128} ] ca_pos( 4 ) produce_pos( 4 )
 logical domain : (iS14{8}, bS15{1 ex 4}, iS16{128000}, iS17{128})
 allocation domain : (iS16{128000}, iS14{8}, iS17{128}, bS15{1 ex 4})
 contiguity: t t t n
  Merge: iS14{8} and bS15{1 ex 4} -> iS36{8 ex ( 8 * 4 )}
  Merge: iS16{128000} and iS17{128} -> iS54{16384000}
  Merge: iS36{8 ex ( 8 * 4 )} and iS54{16384000} -> iS55{131072000 ex ( ( 8 * 4 ) * 16384000 )}
  Split: iS55{131072000 ex ( ( 8 * 4 ) * 16384000 )} by factor 4 -> iS56{32768000 ex ( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) )}, iS57{4}
  Split: iS56{32768000 ex ( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) )} by factor 1 -> iS58{32768000 ex ( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) )}, iUS59{1}
  Split: iS58{32768000 ex ( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) )} by factor 128 -> iblockIdx.x60{256000 ex ( ceilDiv(( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) ), 128) )}, ithreadIdx.x61{128}
 loop domain : (iblockIdx.x60{256000 ex ( ceilDiv(( ceilDiv(( ( 8 * 4 ) * 16384000 ), 4) ), 128) )}, iUS59{1}, iS57{4}, ithreadIdx.x61{128})
T7_l_float[ iblockIdx.x52{1024000}, iUS51{1}, iS49{4}, ithreadIdx.x53{128} ] ca_pos( 2 ) produce_pos( 4 )
 root domain : (iS22{8}rf, iS23{4}rf, iS20{128000}, iS21{128})
  Merge: iS22{8}rf and iS23{4}rf -> iS24{32}rf
 logical domain : (iS24{32}rf, iS20{128000}, iS21{128})
 contiguity: t t t
  Merge: iS20{128000} and iS21{128} -> iS46{16384000}
  Merge: iS24{32}rf and iS46{16384000} -> iS47{524288000}
  Split: iS47{524288000} by factor 4 -> iS48{131072000}, iS49{4}
  Split: iS48{131072000} by factor 1 -> iS50{131072000}, iUS51{1}
  Split: iS50{131072000} by factor 128 -> iblockIdx.x52{1024000}, ithreadIdx.x53{128}
 loop domain : (iblockIdx.x52{1024000}, iUS51{1}, iS49{4}, ithreadIdx.x53{128})
```

The `iS23` ID of the consumer corresponds to the `bS15` ID of the producer. So, the c2p map is necessary unless we change this rule as well. I'm not sure which is better, but I don't have any particular motivation to change it.

Reply:
I'll add this as a code comment.

Reply:
Ah ok, I didn't realize iS23 would not also be an expanded broadcast. This makes sense then. I don't think this motivates making any other changes. Thanks for the clarification.